core/lib/zksync_core/src/api_server/tx_sender/mod.rs

```diff
@@ -875,13 +899,28 @@ impl<G: L1GasPriceProvider> TxSender<G> {
         .into_api_call_result()
     }
 
-    pub fn gas_price(&self) -> u64 {
-        let gas_price = self.0.l1_gas_price_source.estimate_effective_gas_price();
-        let l1_gas_price = (gas_price as f64 * self.0.sender_config.gas_price_scale_factor).round();
-        let (base_fee, _) = derive_base_fee_and_gas_per_pubdata(
-            l1_gas_price as u64,
-            self.0.sender_config.fair_l2_gas_price,
-        );
+    pub async fn gas_price(&self) -> u64 {
+        let fee_model_params = self.0.batch_fee_input_provider.get_batch_fee_input(true);
+
+        let mut connection = self
+            .0
+            .replica_connection_pool
+            .access_storage_tagged("api")
+            .await
+            .unwrap();
+        let block_args = BlockArgs::pending(&mut connection).await;
+        // If protocol version is not present, we'll use the pre-boojum one
+        let protocol_version = connection
+            .blocks_dal()
+            .get_miniblock_protocol_version_id(block_args.resolved_block_number())
+            .await
+            .unwrap()
+            .unwrap_or(ProtocolVersionId::last_pre_boojum());
+        drop(connection);
+
+        let (base_fee, _) =
+            derive_base_fee_and_gas_per_pubdata(fee_model_params, protocol_version.into());
+
```

Comment:
I'd prefer to move it outside and use it as a pure function, how it was done before